File: ios/APIs/YVPHighlightsAPI.swift
import ExpoModulesCore
import YouVersionPlatform

struct YVPHighlightsAPI {
    
    static func createHighlight(bibleId: Int, passageId: String, color: String, promise: Promise) {
        Task {
            do {
                let response = try await YouVersionAPI.Highlights.createHighlight(
                    bibleId: bibleId,
                    passageId: passageId,
                    color: color
                )
                
                promise.resolve(response)
            } catch {
                promise.reject(error)
            }
        }
    }
    
    static func getHighlights(bibleId: Int, passageId: String, promise: Promise) {
        Task {
            do {
                let response = try await YouVersionAPI.Highlights.getHighlights(
                    bibleId: bibleId,
                    passageId: passageId
                )
                
                let records = response.map(HighlightRecord.init)
        
                promise.resolve(records)
            } catch {
                promise.reject(error)
            }
        }
    }
    
    static func updateHighlight(bibleId: Int, passageId: String, color: String, promise: Promise) {
        Task {
            do {
                let response = try await YouVersionAPI.Highlights.updateHighlight(
                    bibleId: bibleId,
                    passageId: passageId,
                    color: color
                )
                
                promise.resolve(response)
            } catch {
                promise.reject(error)
            }
        }
    }
    
    static func deleteHighlight(bibleId: Int, passageId: String, promise: Promise) {
        Task {
            do {
                let response = try await YouVersionAPI.Highlights.deleteHighlight(
                    bibleId: bibleId,
                    passageId: passageId
                )
                
                promise.resolve(response)
            } catch {
                promise.reject(error)
            }
        }
    }
}


internal struct HighlightRecord: Record {
    init() { }
    
    @Field
    var id: String?
    
    @Field
    var bibleId: Int
    
    @Field
    var passageId: String
    
    @Field
    var color: String
    
    @Field
    var userId: String?
    
    @Field
    var createTime: String?
    
    @Field
    var updateTime: String?
    
    init(_ response: HighlightResponse) {
        self.id = response.id
        self.bibleId = response.bibleId
        self.passageId = response.passageId
        self.color = response.color
        self.userId = response.userId
        self.createTime = response.createTime
        self.updateTime = response.updateTime
    }
}

File: android/src/main/java/com/youversion/reactnativesdk/api/YVPHighlightsApi.kt
package com.youversion.reactnativesdk.api

import com.youversion.platform.core.api.YouVersionApi

object YVPHighlightsApi {
    suspend fun createHighlight(bibleId: Int, passageId: String, color: String): Boolean {
        val response = YouVersionApi.highlights.createHighlight(
            versionId = bibleId,
            passageId = passageId,
            color = color
        )

        return response
    }

    suspend fun getHighlights(bibleId: Int, passageId: String): List<HighlightRecord> {
        val response = YouVersionApi.highlights.highlights(
            versionId = bibleId,
            passageId = passageId
        )

        val records = response.map { HighlightRecord(it) }
        return records
    }

    suspend fun updateHighlight(bibleId: Int, passageId: String, color: String): Boolean {
        val response = YouVersionApi.highlights.updateHighlight(
            versionId = bibleId,
            passageId = passageId,
            color = color
        )

        return response
    }

    suspend fun deleteHighlight(bibleId: Int, passageId: String): Boolean {
        val response = YouVersionApi.highlights.deleteHighlight(
            versionId = bibleId,
            passageId = passageId,
        )

        return response
    }
}